File: src/main/java/cz/encircled/eplayer/core/SeriesFinder.java
package cz.encircled.eplayer.core;

import cz.encircled.eplayer.model.MediaType;
import cz.encircled.eplayer.model.SeriesVideo;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.NotNull;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * @author Encircled on 22/09/2014.
 */
public class SeriesFinder {

    private static final Logger log = LogManager.getLogger();

    private static final Pattern seriesPattern = Pattern.compile("(?i).*s[\\d]{1,2}.?e[\\d]{1,2}.*");

    private static final Pattern replacePattern = Pattern.compile("(?i)s[\\d]{1,2}.?e[\\d]{1,2}");

    @NotNull
    private Matcher seriesMatcher = seriesPattern.matcher("");

    @NotNull
    public Map<String, SeriesVideo> findSeries(@NotNull Collection<MediaType> mediaTypes) {
        Map<String, SeriesVideo> result = new HashMap<>();
        mediaTypes.stream().forEach(mediaType -> {
            if (seriesMatcher.reset(mediaType.getName()).matches()) {
                String replacedName = replacePattern.matcher(mediaType.getName()).replaceAll("");
                SeriesVideo seriesVideo = result.get(replacedName);
                if (seriesVideo == null) {
                    log.debug("New series: {}", replacedName);
                    seriesVideo = new SeriesVideo();
                    seriesVideo.setName(replacedName);
                    result.put(replacedName, seriesVideo);
                }
                mediaType.setSeries(true);
                seriesVideo.addMediaType(mediaType);
            }
        });
        return result;
    }

}

File: src/main/java/cz/encircled/eplayer/core/SeriesFinder.kt
package cz.encircled.eplayer.core

import cz.encircled.eplayer.model.MediaFile
import org.apache.logging.log4j.LogManager
import java.io.File

/**
 * @author Encircled on 22/09/2014.
 */
class SeriesFinder {

    private val log = LogManager.getLogger()
    private val seriesPattern = Regex("s[\\d]{1,2}.?e[\\d]{1,2}")

    fun isSeries(name: String): Boolean = seriesPattern.find(name.toLowerCase()) != null

    fun seriesName(media: MediaFile): String = seriesName(File(media.path))

    /**
     * Find series name in the full name
     */
    fun seriesName(media: File): String {
        val i = seriesPattern.find(media.name.toLowerCase())?.range?.first ?: 0
        return if (i > 0) media.name.substring(0, i) else media.parent
    }

    /**
     * Find all series in a folder, return list of paths
     */
    fun findSeriesForName(folder: String, seriesName: String): List<String> =
        File(folder).walk().maxDepth(1)
            .filter { seriesName(it) == seriesName }
            .map { it.path }
            .toList()

}